README
README

This project implements a compression-based distance function [1] in
order to produce a similarity matrix given a list of files in a source
directory. This similarity matrix can be used for clustering etc.

The code is implemented in C++ using boost libraries and OpenMP for
the parallelisation of the distance computations.

This implementation scans the source directory for any files and
computes pair-wise distances (excluding the symmetric
counter-parts). The files can be named in any way. The program
extracts the basename of the file as an identifier.

CONFIGURATION

OpenMP can be enabled using the --enable-openmp flag at the
configuration step.


EXECUTION

The following command-line parameters are provided:

General Configuration:
  --help                produce help message
  --version             show the version

I/O Configuration:
  --result arg (=./results) results directory.
  --source arg (=./sources) source directory containing the files

NCD Configuration:
  --normalise arg (=1)  normalise if the maximum distance > 1

The source directory and the results directory have to be present. The
program only scans files in the source directory. It does not
recursively find the files in any sub-directories of the source
folder.

Only one-directional distances are computed (upper triangular form)
even though the NCD distance computation [2] may not be exactly
equal. However, for large enough sequences this in-equality is
negligible. The distance computation can be normalised with the
"normalise" flag in order for the distances to be strictly below or
equal to 1.


[1] Clustering by compression
    by: R. Cilibrasi, P. M. B. Vitanyi
    Information Theory, IEEE Transactions on, Vol. 51, No. 4. (April
    2005), pp. 1523-1545, doi:10.1109/tit.2005.844059


ChangeLog
* [0.5] - 2015-04-29
** Changed
 - Forked off my canopy clustering project
 - Read text files in a directory


src/main/main.cc
// Copyright (C) 2013, 2015 Dominik Dahlem <Dominik.Dahlem@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#if HAVE_CONFIG_H
# include <config.h>
#endif

#ifndef __STDC_CONSTANT_MACROS
# define __STDC_CONSTANT_MACROS
#endif /* __STDC_CONSTANT_MACROS */

#include <algorithm>
#include <fstream>
#include <iostream>
#include <sstream>

#include <boost/cstdint.hpp>
#include <boost/foreach.hpp>
#include <boost/filesystem.hpp>
#include <boost/algorithm/string/split.hpp>
#include <boost/algorithm/string/trim.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/unordered_map.hpp>
namespace fs = boost::filesystem;

#include <boost/tuple/tuple.hpp>

#include "CL.hh"
namespace cmain = ncd::main;

#include "NCD.hh"

typedef std::pair<std::string, boost::uint32_t> val;
typedef boost::unordered_map<std::string, val> map;
typedef std::vector<boost::tuple<std::string, std::string, double> > distVec;


int main(int argc, char *argv[])
{
  map m_map;

  std::ios_base::sync_with_stdio(false);
  cmain::args_t args;
  cmain::CL cl;

  if (cl.parse(argc, argv, args)) {
    return EXIT_SUCCESS;
  }

  // read the files into memory
  fs::path sourceDir = fs::path(args.source_dir);
  if(!fs::exists(sourceDir) || !fs::is_directory(sourceDir)) {
    std::cerr << "Error: The source directory " << args.source_dir << " needs to exist!" << std::endl;
    return EXIT_FAILURE;
  } else {
    fs::recursive_directory_iterator it(sourceDir);
    fs::recursive_directory_iterator endit;

    while(it != endit) {
      std::ifstream in;
      in.open(it->path().c_str());
      if (!in.is_open()) {
        std::cerr << "Error: Could not open file " << it->path() << std::endl;
        return EXIT_FAILURE;
      } else {
        m_map[it->path().stem().string()] = std::pair<std::string, boost::uint32_t> (
            static_cast<std::stringstream const&>(std::stringstream() << in.rdbuf()).str(), 0);
        in.close();
      }
      ++it;
    }
  }

  // compute the compressed number of bytes for each file in the map
  #pragma omp parallel for shared(m_map) default(none)
  for (boost::uint32_t i = 0; i < m_map.size(); ++i) {
    map::iterator h_i = m_map.begin();
    std::advance(h_i, i);
    boost::uint32_t c = ncd::compressStr((h_i->second).first.c_str(), (h_i->second).first.length());

    #pragma omp critical
    {
      (h_i->second).second = c;
#ifndef NDEBUG
      std::cout << h_i->first << "," << (h_i->second).first.length() << "," << (h_i->second).second << std::endl;
#endif
    }
  }

  // store the distances in a vector
  distVec dists;
  double maxDist = 1.0;

  // compute all distances
  for (boost::uint32_t j = 0; j < m_map.size(); ++j) {
    map::const_iterator it = m_map.begin();
    std::advance(it, j);

    #pragma omp parallel for shared(m_map, std::cout, it, j, dists, maxDist) default(none)
    for (boost::uint32_t i = j+1; i < m_map.size(); ++i) {
      map::const_iterator h_it = m_map.begin();
      std::advance(h_it, i);
      double dist = ncd::NCD<val, double>()(h_it->second, it->second);
      #pragma omp critical
      {
        if (dist > maxDist) {
          maxDist = dist;
        }
        dists.push_back(boost::tuple<std::string, std::string, double>(it->first, h_it->first, dist));
#ifndef NDEBUG
        std::cout << h_it->first << "," << it->first << "," << dist << std::endl;
#endif
      }
    }
  }

  // serialise the similarity matrix
  std::ofstream out;
  std::string outFile = args.results_dir + "/sim.csv";
  out.open(outFile.c_str());
  distVec::iterator it_end = dists.end();
  for (distVec::iterator it = dists.begin(); it != it_end; ++it) {
    double dist = it->get<2>();
    if (args.normalise) {
      dist /= maxDist;
    }
    out << it->get<0>() << "," << it->get<1>() << "," << dist << std::endl;
  }
  out.close();

  return EXIT_SUCCESS;
}


src/main/CL.cc
// Copyright (C) 2013, 2015 Dominik Dahlem <Dominik.Dahlem@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

/** @file CL.cc
 * Implementation of the command-line parsing of the main routine
 *
 * @author Dominik Dahlem
 */
#if HAVE_CONFIG_H
# include <config.h>
#endif

#ifndef __STDC_CONSTANT_MACROS
# define __STDC_CONSTANT_MACROS
#endif /* __STDC_CONSTANT_MACROS */

#include <iostream>

#include <boost/cstdint.hpp>

#include <boost/program_options.hpp>
namespace po = boost::program_options;

#include "CL.hh"



namespace ncd
{
namespace main
{


CL::CL()
    : m_opt_desc(new po::options_description("Options"))
{
  // Declare the supported options.
  po::options_description opt_general("General Configuration");
  opt_general.add_options()
      (HELP.c_str(), "produce help message")
      (VERS.c_str(), "show the version")
      ;

  po::options_description opt_io("I/O Configuration");
  opt_io.add_options()
      (RESULTS_DIR.c_str(), po::value <std::string>()->default_value("./results"), "results directory.")
      (SOURCE_DIR.c_str(), po::value <std::string>()->default_value("./sources"), "source directory containing the files")
      ;

  po::options_description opt_can("NCD Configuration");
  opt_can.add_options()
      (NORMALISE.c_str(), po::value <bool>()->default_value(1), "normalise if the maximum distance > 1")
      ;

  m_opt_desc->add(opt_general);
  m_opt_desc->add(opt_io);
  m_opt_desc->add(opt_can);
}


int CL::parse(int argc, char *argv[], args_t &p_args)
{
  po::variables_map vm;

  po::store(po::parse_command_line(argc, argv, (*m_opt_desc.get())), vm);
  po::notify(vm);

  if (vm.count(HELP)) {
    std::cout << (*m_opt_desc.get()) << std::endl;
    return EXIT_FAILURE;
  }

  if (vm.count(VERS)) {
    std::cout << PACKAGE_NAME << " " << PACKAGE_VERSION << std::endl;
    std::cout << argv[0] << std::endl;
    return EXIT_FAILURE;
  }

  if (vm.count(RESULTS_DIR.c_str())) {
    p_args.results_dir = vm[RESULTS_DIR.c_str()].as <std::string>();
  }

  if (vm.count(SOURCE_DIR.c_str())) {
    p_args.source_dir = vm[SOURCE_DIR.c_str()].as <std::string>();
  }

  if (vm.count(NORMALISE.c_str())) {
    p_args.normalise = vm[NORMALISE.c_str()].as <bool>();
  }

  std::cout << argv[0] << " " << PACKAGE_VERSION << std::endl;
  std::cout << PACKAGE_NAME << std::endl;
  std::cout << p_args << std::endl;

  return EXIT_SUCCESS;
}


}
}
